### PR TITLE
[GEOT-7178] Create a PropertyAccessor that can use XPath to retrieve …

### PR DESCRIPTION
**Pull request by turingtestfail**
[GEOT-7178](https://osgeo-org.atlassian.net/browse/GEOT-7178) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geotools&utm_campaign=badge_info)

…JSONNode attribute children

get stuff

more get

Expanded the tests

cleanup

  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geotools/geotools/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geotools.org/latest/developer/procedures/contribution_license.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] Avoid [Java 9+ split packages](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed).
- [x] All the build checks are green ([see automated QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geotools/geotools/tree/main/docs) has been updated (if change is visible to end users).
- [x] There is an issue in [GeoTools Jira](https://osgeo-org.atlassian.net/projects/GEOT) (except for changes not visible to end users). 
- [x] Commit message(s) must be in the form ``[GEOT-XYZW] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] The commit targets a single objective (if multiple focuses cannot be avoided, each one is in its own commit, and has a separate ticket describing it).

### CODE REVIEW

File: modules/unsupported/geojson-core/src/main/java/org/geotools/filter/expression/JSONNodePropertyAccessorFactory.java [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+                        if (parts.length < 2) {
+                            // only one part means return whole JsonNode
+                            return true;
+                        } else {
+                            JsonPointer pointer = JsonPointer.compile(parts[1]);
+                            JsonNode jsonNode = ((JsonNode) property.getValue()).at(pointer);
+                            return !jsonNode.isMissingNode();
+                        }
+                    }
+                }
```
Comment (aaime) [inline modules/unsupported/geojson-core/src/main/java/org/geotools/filter/expression/JSONNodePropertyAccessorFactory.java:76]: Most of this code is duplicated below. Can you extract a shared method that would extract the target jsonNode or null, and then use it in a different way in canHandle and in get?

File: modules/unsupported/geojson-core/src/main/java/org/geotools/filter/expression/JSONNodePropertyAccessorFactory.java [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+                    Property property = feature.getProperty(parts[0]);
+                    Object value = null;
+                    if (property == null) {
+                        // check to see if it is in the user data map which is sometimes populated
+                        // from geojson
+                        value = getValueFromUserData(parts, feature, value);
+                    } else {
+                        value = property.getValue();
+                    }
+                    if (value != null && value instanceof JsonNode) {
```
Comment (aaime) [inline modules/unsupported/geojson-core/src/main/java/org/geotools/filter/expression/JSONNodePropertyAccessorFactory.java:66]: Does the same:

```suggestion
                    if (value instanceof JsonNode) {
```

File: modules/unsupported/geojson-core/src/main/java/org/geotools/filter/expression/JSONNodePropertyAccessorFactory.java [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+                } else if (jsonNode.isFloatingPointNumber()) {
+                    return jsonNode.doubleValue();
+                } else if (jsonNode.isIntegralNumber()) {
+                    return jsonNode.longValue();
+                } else if (jsonNode.isBigInteger()) {
+                    return jsonNode.bigIntegerValue();
+                } else if (jsonNode.isShort()) {
+                    return jsonNode.shortValue();
+                } else {
+                    return (Double) jsonNode.numberValue();
```
Comment (aaime) [inline modules/unsupported/geojson-core/src/main/java/org/geotools/filter/expression/JSONNodePropertyAccessorFactory.java:177]: Unnecessary cast.

File: modules/unsupported/geojson-core/src/main/java/org/geotools/filter/expression/JSONNodePropertyAccessorFactory.java [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+        private static final String JSON_NODE_DATE_FORMAT_OFF = "JSON_NODE_DATE_FORMAT_OFF";
+
+        @Override
+        public boolean canHandle(Object object, String xpath, Class target) {
+            try {
+                if (xpath != null && object != null && object instanceof SimpleFeature) {
+                    String[] parts = stripAndReturnHeadAndRest(xpath);
+                    SimpleFeature feature = (SimpleFeature) object;
+                    Property property = feature.getProperty(parts[0]);
```
Comment (aaime) [inline modules/unsupported/geojson-core/src/main/java/org/geotools/filter/expression/JSONNodePropertyAccessorFactory.java:57]: May be simpler if you called ``getAttribute`` instead (getting the value directly)

File: modules/unsupported/geojson-core/src/main/java/org/geotools/filter/expression/JSONNodePropertyAccessorFactory.java [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+/** Property Accessor Factory for {@link JsonNode} */
+public class JSONNodePropertyAccessorFactory implements PropertyAccessorFactory {
+
+    static PropertyAccessor JSONNODEPROPERTY = new JSONNodePropertyAccessor();
+    static final String ROOT = "/";
+
+    @Override
+    public PropertyAccessor createPropertyAccessor(
+            Class<?> type, String xpath, Class<?> target, Hints hints) {
+        if (xpath == null) {
```
Comment (aaime) [inline modules/unsupported/geojson-core/src/main/java/org/geotools/filter/expression/JSONNodePropertyAccessorFactory.java:40]: Should you check if "type" is a SimpleFeature? See SimpleFeaturePropertyAccessorFactory.

File: modules/unsupported/geojson-core/src/test/java/org/geotools/filter/expression/JSONNodePropertyAccessorTest.java [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+
+        type = typeBuilder.buildFeatureType();
+
+        SimpleFeatureBuilder builder = new SimpleFeatureBuilder(type);
+
+        builder.add("testString");
+
+        builder.add(new TextNode("testTextNode"));
+
+        ObjectNode threeLevelObject = objectMapper.createObjectNode();
```
Comment (aaime) [inline modules/unsupported/geojson-core/src/test/java/org/geotools/filter/expression/JSONNodePropertyAccessorTest.java:71]: Works, but it's a bit hard to imagine the resulting JSON... Wouldn't it have been easier to write a small JSON snippet, and have ObjectMapper parse it instead?

File: modules/unsupported/geojson-core/src/test/java/org/geotools/filter/expression/JSONNodePropertyAccessorTest.java [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+import org.geotools.feature.simple.SimpleFeatureBuilder;
+import org.geotools.feature.simple.SimpleFeatureTypeBuilder;
+import org.geotools.test.TestData;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.opengis.feature.simple.SimpleFeature;
+import org.opengis.feature.simple.SimpleFeatureType;
+
+public class JSONNodePropertyAccessorTest {
```
Comment (aaime) [inline modules/unsupported/geojson-core/src/test/java/org/geotools/filter/expression/JSONNodePropertyAccessorTest.java:40]: Nice testing, thanks!

**Review by aaime — CHANGES_REQUESTED** [review]

File: modules/unsupported/geojson-core/src/main/java/org/geotools/filter/expression/JSONNodePropertyAccessorFactory.java (hunk 1 above)
Comment (turingtestfail) [reply modules/unsupported/geojson-core/src/main/java/org/geotools/filter/expression/JSONNodePropertyAccessorFactory.java]: Extracted into seperate method

File: modules/unsupported/geojson-core/src/main/java/org/geotools/filter/expression/JSONNodePropertyAccessorFactory.java (hunk 3 above)
Comment (turingtestfail) [reply modules/unsupported/geojson-core/src/main/java/org/geotools/filter/expression/JSONNodePropertyAccessorFactory.java]: fixed

File: modules/unsupported/geojson-core/src/main/java/org/geotools/filter/expression/JSONNodePropertyAccessorFactory.java (hunk 4 above)
Comment (turingtestfail) [reply modules/unsupported/geojson-core/src/main/java/org/geotools/filter/expression/JSONNodePropertyAccessorFactory.java]: changed

File: modules/unsupported/geojson-core/src/main/java/org/geotools/filter/expression/JSONNodePropertyAccessorFactory.java (hunk 5 above)
Comment (turingtestfail) [reply modules/unsupported/geojson-core/src/main/java/org/geotools/filter/expression/JSONNodePropertyAccessorFactory.java:40]: Added

File: modules/unsupported/geojson-core/src/test/java/org/geotools/filter/expression/JSONNodePropertyAccessorTest.java (hunk 6 above)
Comment (turingtestfail) [reply modules/unsupported/geojson-core/src/test/java/org/geotools/filter/expression/JSONNodePropertyAccessorTest.java]: Moved to json in a seperate file as a SimpleFeature for better readability.

**Pull request merged**